middlewares/cors.go

```diff
@@ -0,0 +1,100 @@
+package middlewares
+
+import (
+	"github.com/savsgio/atreugo/v10"
+
+	"errors"
+	"strconv"
+	"strings"
+)
+
+// Create new properties for CORS.
+type CorsOptions struct {
+	// given origin or origins can be shared with response
+	AllowedOrigins []string
+
+	// which headers can be used during the actual request
+	AllowedHeaders []string
+
+	// specify the method or methods allowed to resource
+	AllowedMethods []string
+
+	// which header names can be exposed as part of the response
+	ExposedHeaders []string
+
+	// whether a cached response can be used
+	AllowedVary []string
+
+	// seconds for caching the preflight result
+	AllowMaxAge int
+
+	// whether to expose the response to frontend code
+	AllowCredentials bool
+}
+
+type CorsHandler struct {
+	options CorsOptions
+}
+
+func NewCorsMiddleware(options CorsOptions) atreugo.Middleware {
+	cors := CorsHandler{
+		options: options,
+	}
+
+	return cors.middleware
+}
+
+func (c *CorsHandler) middleware(ctx *atreugo.RequestCtx) error {
+	if err := c.handlePreflight(ctx); err != nil {
+		return err
+	}
+
+	return ctx.Next()
+}
+
+func (c *CorsHandler) handlePreflight(ctx *atreugo.RequestCtx) error {
+	originHeader := string(ctx.Request.Header.Peek("Origin"))
+
+	if !c.isAllowedOrigin(originHeader) {
+		return errors.New("CORS blocked")
```

Comment:
Why??

Reply:
I said you that the locks it's done by the client or browser not by the server.

Reply:
```
	if !c.isAllowedOrigin(originHeader) {
		return ctx.Next()
	}
```
How to write the test when use this?